### PR TITLE
Fix sliproads scenario 4

### PR DESCRIPTION
**Pull request by oxidase**
# Issue

PR fixes scenario 4 of  #4348 by allowing sliproads from links via links.

## Tasklist
 - [x] add regression / cucumber cases (see docs/testing.md)
 - [x] make a map of locations with modified instructions
 - [ ] review
 - [ ] adjust for comments

## Requirements / Relations
 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

**Comment by daniel-j-h** [comment]
Removing the link-class check here looks good but we really need to look at real-world situations to make a call here.

- needs map to inspect situations we now classify as sliproads
- needs update to our maneuver classification paper

**Comment by MoKob** [comment]
@oxidase before reviewing, could you provide the map of `real-world situations` that @daniel-j-h asked for. Without that, the implications of your changes are impossible to assess.
With that map, look at approximately hundred changes and make a note of improved / worsened classification (jugdement call: is this a single turn or is it two).

**Comment by oxidase** [comment]
@MoKob here a map in DE and CA with 671 modified locations https://api.mapbox.com/styles/v1/oxidase/cj8ctmb3z8il92qmnnpzkt3q2.html?fresh=true&title=true&access_token=pk.eyJ1Ijoib3hpZGFzZSIsImEiOiJjaXcxNWMxYmswMDRlMnlxOWtlem0yamU3In0.PjtHAtgQDzxhZ1BE484Sqw#7.19/51.554/6.938 

Most of nodes are like https://www.mapbox.com/get-directions/#17.24/51.54773/7.21275?coordinates=7.211749965246469,51.54783232143697;7.212753812634247,51.54772639593111 
or https://www.mapbox.com/get-directions/#15.68/51.3893/6.2664?coordinates=6.267416425333607,51.387693856479984;6.267469002825692,51.388199698191016 with two instructions.

But in some rare cases https://www.mapbox.com/get-directions/#16.75/37.86651/-122.30449?coordinates=-122.30407045878566,37.86592529445568;-122.30339574159481,37.86677712776516 where a single instruction changed to "Turn right onto University Avenue"

File: src/extractor/guidance/sliproad_handler.cpp [hunk 1]
```diff
@@ -698,24 +698,20 @@ bool SliproadHandler::isValidSliproadArea(const double max_area,
...
-    const auto &first_road_data = node_based_graph.GetEdgeData(first.eid);
-    if (first_road_data.road_classification.IsLinkClass())
-    {
-        return false;
-    }
+    // otherwise the first road leading to the intersection we shortcut
+    // can be a link or a usual road (ignore the check)
 
-    // nor the second road coming from the intersection we shotcut must be links
+    // the second road coming from the intersection we shotcut must be a link
```
Comment (MoKob) [inline src/extractor/guidance/sliproad_handler.cpp]: *shortcut

File: src/extractor/guidance/sliproad_handler.cpp [hunk 2]
```diff
@@ -698,24 +698,20 @@ bool SliproadHandler::isValidSliproadArea(const double max_area,
...
+    // If the sliproad is not a link we don't care
     const auto &sliproad_data = node_based_graph.GetEdgeData(sliproad.eid);
     if (!sliproad_data.road_classification.IsLinkClass())
     {
         return true;
     }
 
-    // otherwise neither the first road leading to the intersection we shortcut
-    const auto &first_road_data = node_based_graph.GetEdgeData(first.eid);
-    if (first_road_data.road_classification.IsLinkClass())
```
Comment (MoKob) [inline src/extractor/guidance/sliproad_handler.cpp]: Have you tried what happens, if you explicitly allow MotorwayLink classes here? Ramps are different than the primary/.. links, right?

**Comment by oxidase** [comment]
@MoKob I have updated the map with a condition `first_road_data.road_classification.IsLinkClass() &&     !first_road_data.road_classification.IsRampClass()` to use only `ramp-> link -> non-link` (red nodes) instead of `any -> link -> non-link` and number of affected intersections halved 
https://api.mapbox.com/styles/v1/oxidase/cj8ctmb3z8il92qmnnpzkt3q2.html?title=true&access_token=pk.eyJ1Ijoib3hpZGFzZSIsImEiOiJjaXcxNWMxYmswMDRlMnlxOWtlem0yamU3In0.PjtHAtgQDzxhZ1BE484Sqw#18.27/49.46387/8.50602

A typical example is https://www.mapbox.com/get-directions/#18.27/49.46407/8.50528?coordinates=8.50620676926198,49.464227828009285;8.505281415385014,49.46406952432184
with a `primary link -> primary link -> primary` sliproad

**Comment by MoKob** [comment]
@oxidase in that case, we can keep it with the complete removal. It looks like we didn't regress anywhere (or at least in no location I've seen). Good to go from my point of view.

**Pull request merged**